### ISSUE TITLE
Docs - Choose your own adventure Cookbook has an error in the code #473

### ISSUE THREAD

**Issue by nehiljain** [DAGWorks-Inc/burr#474]
[Choose your own adventure cookbook](https://burr.dagworks.io/examples/simple/choose-your-own-adventure/) is not running as expected.

# Current behavior
1. When running the cookbook, I get an error.

## Stack Traces

```
Traceback (most recent call last):
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 885, in _step
    result, new_state = _run_single_step_action(
                        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 287, in _run_single_step_action
    action.run_and_update(state, **inputs), action.name, action.schema
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/action.py", line 655, in run_and_update
    return self._fn(state, **self._bound_params, **run_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/projects/demodrive-tests/burr/run_choose_adventure.py", line 13, in start
    current_challenge = state["challenges"][0]
                        ~~~~~^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/state.py", line 432, in __getitem__
    raise KeyError(
KeyError: 'Key "challenges" not found in state. Keys state knows about are: [\'did_win\']. If you hit this within the context of an application, you want to (a) ensure that an upstream action has produced this state/it is set as an initial state value and (b) ensure that your action declares this as a read key.'

Traceback (most recent call last):
  File "/Users/spalanimalai/projects/demodrive-tests/burr/run_choose_adventure.py", line 143, in <module>
    action, state, result = app.run(halt_after=["win"])
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/telemetry.py", line 276, in wrapped_fn
    return call_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 672, in wrapper_sync
    return fn(app_self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 1243, in run
    next(gen)
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 1186, in iterate
    prior_action, result, state = self.step(inputs=inputs)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 672, in wrapper_sync
    return fn(app_self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 843, in step
    out = self._step(inputs=inputs, _run_hooks=True)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 899, in _step
    raise e
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 885, in _step
    result, new_state = _run_single_step_action(
                        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/application.py", line 287, in _run_single_step_action
    action.run_and_update(state, **inputs), action.name, action.schema
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/action.py", line 655, in run_and_update
    return self._fn(state, **self._bound_params, **run_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spalanimalai/projects/demodrive-tests/burr/run_choose_adventure.py", line 13, in start
    current_challenge = state["challenges"][0]
                        ~~~~~^^^^^^^^^^^^^^
  File "/Users/spalanimalai/.pyenv/versions/pretzelai/lib/python3.12/site-packages/burr/core/state.py", line 432, in __getitem__
    raise KeyError(
KeyError: 'Key "challenges" not found in state. Keys state knows about are: [\'did_win\']. If you hit this within the context of an application, you want to (a) ensure that an upstream action has produced this state/it is set as an initial state value and (b) ensure that your action declares this as a read key.'
```

## Recommended Fix
Pass additional state. `.with_state(did_win=False, challenges=challenges, current_challenge=challenges[0])`

# Additional context
This bug was found using an automation built by https://www.demodrive.tech/. We are looking at popular OSS projects and helping them ensure their documentation is error-proof.

**Comment by skrawcz** [comment]
@nehiljain thanks for flagging I think I have a fix - #475

**Comment by skrawcz** [comment]
@nehiljain please try again -- let me know if it doesn't work.

**Comment by nehiljain** [comment]
> @nehiljain thanks for flagging I think I have a fix - #475

@skrawcz does this fix cowsay or choose your own adventure? I think you meant to mark #473 closed instead of this one.

**Comment by skrawcz** [comment]
@nehiljain yep my bad. But I have fix for that too! Thanks for the finds. Interesting tool. Will DM.
